### DISCUSSION TITLE
Loading page before the DOM loads

### DISCUSSION THREAD

**Discussion by mares-gabriel** [vercel/next.js#52217] [Help]
### Summary

I am trying to add a custom loading page, before the DOM loads, but some for reason, it does not work. I tried to add an useEffect in the layout.tsx file, that has an event listener on 'DOMContentLoaded' or 'load', but it does not work. I want to show this particular loading screen only when the user enters the first time on the website.

**Comment by rasif-sahl** [comment]
Create a new component for your custom loading page. This component will be responsible for displaying the loading screen.

```
// CustomLoading.js

import { useEffect, useState } from 'react';

function CustomLoading() {
  const [showLoading, setShowLoading] = useState(true);

  useEffect(() => {
    // Check if the user has visited the website before
    const hasVisited = localStorage.getItem('hasVisited');
    if (hasVisited) {
      setShowLoading(false);
    } else {
      localStorage.setItem('hasVisited', 'true');
      // Perform any additional actions for the first visit
      // For example, you can fetch initial data here
      // Once the data is fetched, you can update the loading state and proceed to display the content
      fetchData().then(() => {
        setShowLoading(false);
      });
    }
  }, []);

  if (showLoading) {
    return <div>Loading...</div>; // Replace with your custom loading screen
  }

  return null; // Render nothing when loading is complete
}

export default CustomLoading;
```

In the CustomLoading component, an useEffect hook is used to check if the user has visited the website before by checking the value of the hasVisited key in localStorage. If the user has visited before, the loading state is set to false, and the loading screen is not displayed. Otherwise, for the first visit, additional actions can be performed (e.g., fetching initial data), and once those actions are completed, the loading state is set to false.

2. Import and use the CustomLoading component in your layout.tsx file or any other appropriate component that wraps the content of your pages.

```
// layout.tsx

import CustomLoading from './CustomLoading';

function Layout({ children }) {
  return (
    <div>
      <CustomLoading />
      {/* Render other layout components and the content */}
      {children}
    </div>
  );
}

export default Layout;
```

By adding the CustomLoading component to your layout, it will be displayed before the DOM loads and only for the user's first visit to the website. Once the loading is complete, the content will be rendered.

Make sure to replace <div>Loading...</div> in the CustomLoading component with your actual custom loading screen or component.

Remember to adjust the logic and behavior according to your specific requirements.

**Reply by dinesh-hills** [reply]
Exactly what I was looking for, thanks @rasif-sahl

**Reply by rasif-sahl** [reply]
great...

**Comment by masterbater** [comment]
Turns out you can use template special file for nextjs where every nagivation it will show the loading fallback. 
https://nextjs.org/docs/app/building-your-application/routing/pages-and-layouts#templates
But your concern is showing a loading spinner while resources are being fetch or doing a request, I believe this is handled by the root loading special file.
